1. 
[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin:1.5.3.RELEASE:run (default-cli) on project datariver: Unable to find a suitable main class, please add a 'mainClass' property
 Add
@SpringBootApplication
public class DatariverApplication {
	public static in main(String[] args) {
		SpringApplication.run(DatariverApplication.class, args);
	}
}


2.
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaAutoConfiguration.class]: Invocation of init method failed; nested exception is org.hibernate.HibernateException: Access to DialectResolutionInfo cannot be null when 'hibernate.dialect' not set

add following properties in the application.properties, the problem still reported.
spring.datasource.hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
spring.jpa.hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
spring.hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
hibernate.dialect=org.hibernate.dialect.MySQL5Dialect


spring.datasource.url=jdbc\:mysql\://localhost\:3306/xe?useUnicode\=true\&characterEncoding\=UTF-8\&autoReconnect\=true\&failOverReadOnly\=false

the actual cause is there are no database named "xe".


3.
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dataSource' defined in class path resource [org/springframework/boot/autoconfigure/jdbc/DataSourceConfiguration$Tomcat.class]: Initialization of bean failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dataSourceInitializer': Invocation of init method failed; nested exception is org.springframework.jdbc.datasource.init.UncategorizedScriptException: Failed to execute database script; nested exception is org.springframework.jdbc.CannotGetJdbcConnectionException: Could not get JDBC Connection; nested exception is com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Could not create connection to database server. 

schema.sql or data.sql has syntax error 
And attention, "CREATE DATABASE  IF NOT EXISTS datarivers;" is meaningless.  For setup a database, it must specified a database, and only
operate this database. So consider that you can not create a database in schema.sql or data.sql.

mysql> source /home/zen/Work/ShineWebService/src/main/resources/schema.sql
you should source sql script twice, for some failure occurs on update or drop table when the table has foreign key.


Caused by: org.springframework.jdbc.datasource.init.ScriptStatementFailedException: Failed to execute SQL script statement #3 of URL [file:/home/zen/Work/ShineWebService/target/classes/schema.sql]: DROP TABLE IF EXISTS role; nested exception is com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Cannot delete or update a parent row: a foreign key constraint fails


4.
Consider defining a bean of type 'org.springframework.security.core.userdetails.UserDetailsService' 

add @Component for UserDetailsServiceImpl. Or this set in .xml files
@Component
public class UserDetailsServiceImpl implements UserDetailsService {
}


For jetty-maven-plugin , xml configuration locates in path webapp/WEB-INF/web.xml

2017-04-30 10:08:04.670  INFO 8084 --- [lication.main()] j.LocalContainerEntityManagerFactoryBean : Building JPA container EntityManagerFactory for persistence unit 'default'
2017-04-30 10:08:04.690  INFO 8084 --- [lication.main()] o.hibernate.jpa.internal.util.LogHelper  : HHH000204: Processing PersistenceUnitInfo [
	name: default
	...]
2017-04-30 10:08:04.891  INFO 8084 --- [lication.main()] org.hibernate.Version                    : HHH000412: Hibernate Core {4.3.10.Final}
2017-04-30 10:08:04.904  INFO 8084 --- [lication.main()] org.hibernate.cfg.Environment            : HHH000206: hibernate.properties not found
2017-04-30 10:08:04.908  INFO 8084 --- [lication.main()] org.hibernate.cfg.Environment            : HHH000021: Bytecode provider name : javassist
2017-04-30 10:08:05.433  INFO 8084 --- [lication.main()] o.hibernate.annotations.common.Version   : HCANN000001: Hibernate Commons Annotations {4.0.5.Final}
2017-04-30 10:08:06.479  INFO 8084 --- [lication.main()] org.hibernate.dialect.Dialect            : HHH000400: Using dialect: org.hibernate.dialect.MySQL5Dialect
2017-04-30 10:08:06.927  INFO 8084 --- [lication.main()] o.h.h.i.ast.ASTQueryTranslatorFactory    : HHH000397: Using ASTQueryTranslatorFactory
2017-04-30 10:08:07.647  INFO 8084 --- [lication.main()] org.hibernate.tool.hbm2ddl.SchemaUpdate  : HHH000228: Running hbm2ddl schema update
2017-04-30 10:08:07.647  INFO 8084 --- [lication.main()] org.hibernate.tool.hbm2ddl.SchemaUpdate  : HHH000102: Fetching database metadata
2017-04-30 10:08:07.651  INFO 8084 --- [lication.main()] org.hibernate.tool.hbm2ddl.SchemaUpdate  : HHH000396: Updating schema
2017-04-30 10:08:07.673  INFO 8084 --- [lication.main()] java.sql.DatabaseMetaData                : HHH000262: Table not found: person
2017-04-30 10:08:07.678  INFO 8084 --- [lication.main()] java.sql.DatabaseMetaData                : HHH000262: Table not found: person
2017-04-30 10:08:07.684  INFO 8084 --- [lication.main()] java.sql.DatabaseMetaData                : HHH000262: Table not found: person
2017-04-30 10:08:07.947  INFO 8084 --- [lication.main()] org.hibernate.tool.hbm2ddl.SchemaUpdate  : HHH000232: Schema update complete
2017-04-30 10:08:08.596  INFO 8084 --- [lication.main()] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from URL [file:/home/zen/Spring/springboot-wyf-sourcecode/ch8_2/src/main/resources/data.sql]
2017-04-30 10:08:08.709  WARN 8084 --- [lication.main()] ationConfigEmbeddedWebApplicationContext : Exception encountered during context initialization - cancelling refresh attempt




curl -v -H "Content-Type: application/json" -d "{ \"username\": \"danveloper\", \"firstName\": \"Dan\", \"lastName\": \"Woods\", \"createdDate\": \"2014-02-02T00:00:00\" }" http://localhost:8080/user

5.
Could not open Hibernate Session for transaction; nested exception is java.lang.NoClassDefFoundError: org/hibernate/engine/transaction/spi/TransactionContext
导入的hibernate的包是5.0的， org/hibernate/engine/transaction/spi/TransactionContext这个类已经移除了


/home/zen/.m2/repository/org/hibernate/hibernate-core/5.0.12.Final/hibernate-core-5.0.12.Final.jar
/home/zen/.m2/repository/org/hibernate/hibernate-core/4.3.10.Final/hibernate-core-4.3.10.Final.jar
org/hibernate/engine/transaction/spi/TransactionContext.class
/home/zen/.m2/repository/org/hibernate/hibernate-core/4.3.11.Final/hibernate-core-4.3.11.Final.jar
org/hibernate/engine/transaction/spi/TransactionContext.class

mvn dependency:tree
+- org.springframework.boot:spring-boot-starter-data-jpa:jar:1.5.3.RELEASE:compile
|  +- org.hibernate:hibernate-core:jar:5.0.12.Final:compile
|  +- org.hibernate:hibernate-entitymanager:jar:5.0.12.Final:compile
mvn dependency:list


解决方案： 把spring-boot 从1.5.3.Release 降到 1.3.0.Release。


6.
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'springSecurityFilterChain' defined in class path resource [org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.servlet.Filter]: Factory method 'springSecurityFilterChain' threw exception; nested exception is java.lang.IllegalArgumentException: ROLE_ADMIN cannot start with ROLE_ (it is automatically added)

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'springSecurityFilterChain' defined in class path resource [org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.servlet.Filter]: Factory method 'springSecurityFilterChain' threw exception; nested exception is org.springframework.security.config.annotation.AlreadyBuiltException: This object has already been built

the right name:
antMatchers("/", "/welcome").access("hasRole('USER') or hasRole('ADMIN') or hasRole('DBA')")
the error role, and cause the above error:
.antMatchers("/", "/welcome").access("hasRole('ROLE_USER') or hasRole('ROLE_ADMIN') or hasRole('ROLE_DBA')")
auth.inMemoryAuthentication()
.withUser("hill").password("asd@MBB").roles("ROLE_ADMIN");

7.
spring.datasource.driverClassName=org.postgresql.Driver
spring.datasource.url=jdbc:postgresql://localhost:5432/teste?charSet=LATIN1
spring.datasource.username=klebermo
spring.datasource.password=123

spring.jpa.database-platform=org.hibernate.dialect.PostgreSQLDialect
spring.jpa.show-sql=false
spring.jpa.hibernate.ddl-auto=create


 HHH000342: Could not obtain connection to query metadata : ORA-01017: invalid username/password; logon denied

 Exception encountered during context initialization - cancelling refresh attempt

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaAutoConfiguration.class]: Invocation of init method failed; nested exception is org.hibernate.HibernateException: Access to DialectResolutionInfo cannot be null when 'hibernate.dialect' not set

java.sql.SQLException: ORA-01017: invalid username/password; logon denied
用户不存在或密码错误

java.sql.SQLException: ORA-01045: user BOOT lacks CREATE SESSION privilege; logon denied
权限不够。

	at oracle.jdbc.driver.T4CTTIoer.processError(T4CTTIoer.java:447)
	at oracle.jdbc.driver.T4CTTIoer.processError(T4CTTIoer.java:389)
	at oracle.jdbc.driver.T4CTTIoer.processError(T4CTTIoer.java:382)
	at oracle.jdbc.driver.T4CTTIfun.processError(T4CTTIfun.java:675)
	at oracle.jdbc.driver.T4CTTIoauthenticate.processError(T4CTTIoauthenticate.java:448)
	at oracle.jdbc.driver.T4CTTIfun.receive(T4CTTIfun.java:513)
	at oracle.jdbc.driver.T4CTTIfun.doRPC(T4CTTIfun.java:227)
	at oracle.jdbc.driver.T4CTTIoauthenticate.doOAUTH(T4CTTIoauthenticate.java:383)
	at oracle.jdbc.driver.T4CTTIoauthenticate.doOAUTH(T4CTTIoauthenticate.java:776)
	at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:432)
	at oracle.jdbc.driver.PhysicalConnection.<init>(PhysicalConnection.java:553)
	at oracle.jdbc.driver.T4CConnection.<init>(T4CConnection.java:254)
	at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:32)
	at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:528)




7.
@Resetcontrol 用于返回对象，会自动格式化为JSON
@Control会返回jsp, html页面


8.
如果jsp页面没有被解析，那么需要申明库jstl 以及 tomcat-embed-jasper，
jstl不使用会报错，但是tomcat-embed-jasper不使用，不会报错，但是jsp不会解析，只
是原文返回给浏览器：

        <!--配置jsp jstl的支持-->  
        <dependency>  
            <groupId>javax.servlet</groupId>  
            <artifactId>jstl</artifactId>  
			<version>${jstl.version}</version>
        </dependency>  
  
        <!--对jsp的支持-->  
        <dependency>  
            <groupId>org.apache.tomcat.embed</groupId>  
            <artifactId>tomcat-embed-jasper</artifactId>  
        </dependency>  


        <!--添加对tomcat的支持-->  
        <dependency>  
           <groupId>org.springframework.boot</groupId>  
           <artifactId>spring-boot-starter-tomcat</artifactId>  
            <scope>provided</scope>  
        </dependency>  
		
        <!--配置servlet-->  
        <dependency>  
            <groupId>javax.servlet</groupId>  
            <artifactId>javax.servlet-api</artifactId>  
        </dependency>  
